axelrod/player.py

```diff
@@ -122,6 +125,44 @@ def __init__(self):
         self.state_distribution = defaultdict(int)
         self.set_match_attributes()
 
+
+    def __eq__(self, other):
+        """
+        Test if two players are equal.
+        """
+        if self.__repr__() != other.__repr__():
+            return False
+        for attribute in set(list(self.__dict__.keys()) +
+                             list(other.__dict__.keys())):
+
+            value = getattr(self, attribute, None)
+            other_value = getattr(other, attribute, None)
+
+            if isinstance(value, np.ndarray):
+                if not (np.array_equal(value, other_value)):
+                    return False
+
+            elif isinstance(value, types.GeneratorType) or \
+                 isinstance(value, itertools.cycle):
+
+                # Split the original generator so it is not touched
+                generator, original_value = itertools.tee(value)
+                other_generator, original_other_value = itertools.tee(other_value)
+
+                setattr(self, attribute,
+                        (ele for ele in original_value))
+                setattr(other, attribute,
+                        (ele for ele in original_other_value))
+
+                if not (all(next(generator) == next(other_generator)
```

Comment:
FYI `test_equality_on_init` is the test that fails in this case.

This should clarify why:

```python
>>> import itertools
>>> import types
>>> cycle = itertools.cycle('ABCD')
>>> new_cycle, backup = itertools.tee(cycle)
>>> type(cycle), type(new_cycle), type(backup)
(itertools.cycle, itertools._tee, itertools._tee)
>>> isinstance(cycle, itertools.cycle), isinstance(cycle, types.GeneratorType)
(True, False)
>>> isinstance(new_cycle, itertools.cycle), isinstance(backup, types.GeneratorType)
(False, False)
```

So the second time around the generator/cycle is not compared in the correct way. I did have a special case for a cycle but found it more elegant to just treat them all the same way. :)